src/Symfony/Component/Translation/Provider/ProvidersFactory.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Provider;
+
+use Symfony\Component\Translation\Exception\UnsupportedSchemeException;
+
+/**
+ * @author Mathieu Santostefano <msantostefano@protonmail.com>
+ *
+ * @experimental in 5.3
+ */
+class ProvidersFactory
+{
+    private $factories;
+    private $enabledLocales;
+
+    /**
+     * @param ProviderFactoryInterface[] $factories
+     */
+    public function __construct(iterable $factories, array $enabledLocales)
+    {
+        $this->factories = $factories;
+        $this->enabledLocales = $enabledLocales;
+    }
+
+    public function fromConfig(array $config): TranslationProviders
+    {
+        $providers = [];
+        foreach ($config as $name => $currentConfig) {
+            $providers[$name] = $this->fromString(
+                $currentConfig['dsn'],
+                !$currentConfig['locales'] ? $this->enabledLocales : $currentConfig['locales'],
+                !$currentConfig['domains'] ? [] : $currentConfig['domains']
+            );
+        }
+
+        return new TranslationProviders($providers);
+    }
+
+    public function fromString(string $dsn, array $locales, array $domains = []): ProviderInterface
+    {
+        return $this->fromDsnObject(Dsn::fromString($dsn), $locales, $domains);
+    }
+
+    public function fromDsnObject(Dsn $dsn, array $locales, array $domains = []): ProviderInterface
+    {
+        foreach ($this->factories as $factory) {
+            if ($factory->supports($dsn)) {
+                return new ProviderDecorator($factory->create($dsn), $locales, $domains);
```

Comment:
Why do we need the ProviderDecorator?

Reply:
The only aim of the ProviderDecorator is to intersect the domains from the provider configuration and the domains passed via the command option. It was designed by @odolbeau, and I hasn't cons against it. May be there is a better way to do?

Reply:
So you basically want to filter the domains and locales. That make sense. 

Maybe rename the class and we should definitly add a class comment.